File: functional/Erlang/caesar_cipher.erl
-module(prog).
-export([shifter/2, encrypt/2, decrypt/2, solve/2, main/0]).


shifter(Char, Shift) when (Char /= " ") ->
                    % if a char, then shift
	                Ascii = fun([Ch]) -> Ch end,
	                (((Ascii(Char) - 65 + Shift) rem (26) + 26) rem 26) + 65.
shifter(Char, _Shift) ->
                    % otherwise, whitespace
                    Char.


encrypt(ToCipher, Shift) ->
    % functionally map each character to shifter method
    Ciphered = lists:map(fun(Char) -> shifter([string:to_upper(Char)], Shift) end, ToCipher),
    Ciphered.

decrypt(Ciphered, Shift) ->
    Ciphered = encrypt(Ciphered, -Shift).

solve(Cipher, 0) ->
    io:format("~s~n", [string:concat("Caesar 0: ",string:to_upper(Cipher))]);
solve(Cipher, Shift) when Shift > 0 -> 
	io:format("~s~n", [lists:concat(["Caesar ", Shift, ": ", encrypt(Cipher, Shift)])]),
	solve(Cipher, (Shift-1)).


%% Main Function
main() ->
    % Initialization
    CipherStr = "Dude Wheres My Car",
	Shift = 4,

    io:fwrite("Welcome to Caesar Ciphers with Erlang!\n"),

    % Call Caesar Cipher Encryption Method
    encrypt(CipherStr, Shift),

    % Call Caesar Cipher Decryption Method
	decrypt(CipherStr, Shift),
	
    % Call Caesar Cipher Solve Method
    solve("DUDE", 26).

File: LanguageDesignProject/PastPHP/caesar_cipher.php
<?PastPHP
<?module(CaesarCipher)
<?functions([encrypt/2, decrypt/2])

function encrypt($toCipher: Str): Str
    <var
        $i: Integer;
        $cipher: Str;
    >
    <begin
        for ($i := 0; $i < $toCipher.length; $i++) do
            // Account for uppercase letters
            'A'..'Z': cipher[i] := chr(ord('A') + (ord(toCipher[i]) - ord('A') + shift) mod 26);
            // Account for lowercase letters
            'a'..'z': cipher[i] := chr(ord('a') + (ord(toCipher[i]) - ord('a') + shift) mod 26);
        done
    end>
    <return $cipher>

function decrypt($ciphered: Str): Str
    <var
        $cipher: Str;
    >
    <begin
        $cipher := encrypt($ciphered, ($shift * -1));
    end>
    <return $cipher>

<past
    <var
        // Global Declarations
        $cipherStr: Str;
        $shift: Int;
    >
    <begin
        // Initialization
        $cipherStr := "If my calculations are correct \n
        when this baby hits eighty eight miles per hour \n
        youre gonna see some serious shiii";
        $shift := 4;

        writeLn("Welcome to a Caesar Cipher in PastPHP!");

        // Call Caesar Cipher Encryption Method
        cipherStr := encrypt(cipherStr, shift);
        Writeln('Encrypted Cipher Value: ' + cipherStr);
        
        // Call Caesar Cipher Decryption Method
        cipherStr := decrypt(cipherStr, shift);
        Writeln('Decrypted Cipher Value: ' + cipherStr);
    end>
future>
?>




